File: dbs5/php/CdmDataClient.php
<?php

// Minimal CDM Data API client. Uses ext-curl (bundled with PHP), no Composer needed.

class CdmDataException extends Exception
{
    public $statusCode;
    public $body;

    public function __construct($message, $statusCode = null, $body = null)
    {
        parent::__construct($message);
        $this->statusCode = $statusCode;
        $this->body = $body;
    }
}

class CdmDataClient
{
    private $accessKey;
    private $secretKey;
    private $baseUrl;

    public function __construct(string $accessKey, string $secretKey, string $baseUrl = 'https://dbs5.cplservice.com')
    {
        $this->accessKey = $accessKey;
        $this->secretKey = $secretKey;
        $this->baseUrl = rtrim($baseUrl, '/');
    }

    // Pure URL building, kept separate so it can be self-checked without curl/network.
    public static function buildUrl(string $baseUrl, string $path, array $params = []): string
    {
        $url = rtrim($baseUrl, '/') . $path;
        $params = array_filter($params, function ($v) {
            return $v !== null;
        });
        if ($params) {
            $url .= '?' . http_build_query($params);
        }
        return $url;
    }

    private function request(string $method, string $path, array $params = [], ?array $jsonBody = null): array
    {
        $url = self::buildUrl($this->baseUrl, $path, $params);

        $headers = [
            'x-api-access-key: ' . $this->accessKey,
            'x-api-secret-key: ' . $this->secretKey,
        ];

        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);

        if ($jsonBody !== null) {
            $headers[] = 'content-type: application/json';
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($jsonBody));
        }

        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

        $response = curl_exec($ch);
        if ($response === false) {
            $error = curl_error($ch);
            curl_close($ch);
            throw new CdmDataException("cURL error: $error");
        }
        $statusCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        $data = json_decode($response, true);
        if ($statusCode >= 400) {
            throw new CdmDataException("HTTP $statusCode", $statusCode, $data);
        }
        return $data;
    }

    public function getData(string $deviceId, array $params = []): array
    {
        return $this->request('GET', '/cdmdata/v1/getdata', ['device_id' => $deviceId] + $params);
    }

    public function getDataHourly(string $deviceId, array $params = []): array
    {
        return $this->request('GET', '/cdmdata/v1/getdatahourly', ['device_id' => $deviceId] + $params);
    }

    public function getDataDaily(string $deviceId, array $params = []): array
    {
        return $this->request('GET', '/cdmdata/v1/getdatadaily', ['device_id' => $deviceId] + $params);
    }

    public function onOffCommand(string $deviceId, string $meterId, string $command): array
    {
        if (!in_array($command, ['on', 'off'], true)) {
            throw new InvalidArgumentException("command must be 'on' or 'off'");
        }
        return $this->request('POST', '/cdmdata/v1/on-off-command', [], [
            'device_id' => $deviceId,
            'meter_id' => $meterId,
            'command' => $command,
        ]);
    }
}

File: dbs5/csharp/CdmDataClient.cs
using System.Text;
using System.Text.Json;

namespace CdmDataSdk;

// Minimal CDM Data API client. Uses only BCL types (HttpClient, System.Text.Json),
// no NuGet package needed.

public class CdmDataException : Exception
{
    public int? StatusCode { get; }
    public string? Body { get; }

    public CdmDataException(string message, int? statusCode = null, string? body = null) : base(message)
    {
        StatusCode = statusCode;
        Body = body;
    }
}

public class CdmDataClient
{
    private readonly HttpClient _http;
    private readonly string _accessKey;
    private readonly string _secretKey;
    private readonly string _baseUrl;

    public CdmDataClient(string accessKey, string secretKey, string baseUrl = "https://dbs5.cplservice.com", HttpClient? httpClient = null)
    {
        _accessKey = accessKey;
        _secretKey = secretKey;
        _baseUrl = baseUrl.TrimEnd('/');
        _http = httpClient ?? new HttpClient();
    }

    // Pure URL building, kept separate so it can be self-checked without HTTP mocking.
    public static string BuildQueryUrl(string baseUrl, string path, Dictionary<string, string?>? query = null)
    {
        var url = baseUrl.TrimEnd('/') + path;
        if (query is { Count: > 0 })
        {
            var pairs = query
                .Where(kv => kv.Value is not null)
                .Select(kv => $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value!)}");
            var qs = string.Join("&", pairs);
            if (qs.Length > 0) url += "?" + qs;
        }
        return url;
    }

    private async Task<JsonDocument> RequestAsync(HttpMethod method, string path, Dictionary<string, string?>? query = null, object? jsonBody = null)
    {
        var url = BuildQueryUrl(_baseUrl, path, query);

        using var req = new HttpRequestMessage(method, url);
        req.Headers.Add("x-api-access-key", _accessKey);
        req.Headers.Add("x-api-secret-key", _secretKey);

        if (jsonBody is not null)
        {
            var json = JsonSerializer.Serialize(jsonBody);
            req.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        using var res = await _http.SendAsync(req);
        var text = await res.Content.ReadAsStringAsync();
        if (!res.IsSuccessStatusCode)
        {
            throw new CdmDataException($"HTTP {(int)res.StatusCode}", (int)res.StatusCode, text);
        }
        return JsonDocument.Parse(text);
    }

    public Task<JsonDocument> GetDataAsync(string deviceId, Dictionary<string, string?>? extraParams = null)
        => RequestAsync(HttpMethod.Get, "/cdmdata/v1/getdata", MergeDeviceId(deviceId, extraParams));

    public Task<JsonDocument> GetDataHourlyAsync(string deviceId, Dictionary<string, string?>? extraParams = null)
        => RequestAsync(HttpMethod.Get, "/cdmdata/v1/getdatahourly", MergeDeviceId(deviceId, extraParams));

    public Task<JsonDocument> GetDataDailyAsync(string deviceId, Dictionary<string, string?>? extraParams = null)
        => RequestAsync(HttpMethod.Get, "/cdmdata/v1/getdatadaily", MergeDeviceId(deviceId, extraParams));

    public Task<JsonDocument> OnOffCommandAsync(string deviceId, string meterId, string command)
    {
        if (command != "on" && command != "off")
            throw new ArgumentException("command must be 'on' or 'off'");

        return RequestAsync(HttpMethod.Post, "/cdmdata/v1/on-off-command", jsonBody: new
        {
            device_id = deviceId,
            meter_id = meterId,
            command
        });
    }

    private static Dictionary<string, string?> MergeDeviceId(string deviceId, Dictionary<string, string?>? extra)
    {
        var result = new Dictionary<string, string?> { ["device_id"] = deviceId };
        if (extra != null)
        {
            foreach (var kv in extra) result[kv.Key] = kv.Value;
        }
        return result;
    }
}
